File: car_factory_app/lib/features/trades/trade_api.dart
import 'package:cloud_functions/cloud_functions.dart';

import '../../core/firebase/firebase_bootstrap.dart';
import '../../shared/models/trade.dart';

/// 직거래 Callable — 웹 `features/trades/trade-api.ts` 대응.
/// 주문 상태 쓰기는 절대 클라이언트에서 직접 하지 않는다.
class TradeApi {
  TradeApi({FirebaseFunctions? functions})
      : _functions = functions ?? FirebaseBootstrap.functions;

  final FirebaseFunctions _functions;

  Future<({String orderId, String orderNumber})> createDirectTrade({
    required String productId,
    required DealMethod dealMethod,
    String? buyerUid,
  }) async {
    final callable = _functions.httpsCallable('createDirectTrade');
    final result = await callable.call<Map<String, dynamic>>({
      'productId': productId,
      'dealMethod': dealMethod == DealMethod.transfer ? 'transfer' : 'meet',
      'buyerUid': ?buyerUid,
    });
    final data = Map<String, dynamic>.from(result.data as Map);
    final orderId = data['orderId'] as String?;
    final orderNumber = data['orderNumber'] as String?;
    if (orderId == null || orderNumber == null) {
      throw StateError('거래를 만들지 못했습니다.');
    }
    return (orderId: orderId, orderNumber: orderNumber);
  }

  Future<void> cancelDirectTrade({
    required String orderId,
    String? cancelReason,
  }) async {
    final callable = _functions.httpsCallable('cancelDirectTrade');
    await callable.call({
      'orderId': orderId,
      'cancelReason': ?cancelReason,
    });
  }

  Future<void> completeDirectTrade({
    required String orderId,
    required FulfillmentMethod fulfillmentMethod,
    Map<String, String>? shippingInfo,
  }) async {
    final callable = _functions.httpsCallable('completeDirectTrade');
    await callable.call({
      'orderId': orderId,
      'fulfillmentMethod':
          fulfillmentMethod == FulfillmentMethod.delivery ? 'delivery' : 'meetup',
      'shippingInfo': ?shippingInfo,
    });
  }
}

File: car_factory_web/src/features/trades/trade-api.ts
import { httpsCallable } from "firebase/functions";
import { getClientFunctions, isFirebaseConfigured } from "@/lib/firebase/client";
import type {
  DealMethod,
  FulfillmentMethod,
  TradeShippingInfo,
} from "@/types/trade";

function stripFirebaseMessage(e: unknown): string {
  const message =
    e &&
    typeof e === "object" &&
    "message" in e &&
    typeof (e as { message: unknown }).message === "string"
      ? (e as { message: string }).message
      : e instanceof Error
        ? e.message
        : "요청에 실패했습니다.";
  return message.replace(/^Firebase:\s*/i, "").replace(/\s*\(.*\)$/, "");
}

export async function createDirectTrade(input: {
  productId: string;
  dealMethod: DealMethod;
  /** 판매자가 채팅 상대(구매자)를 예약할 때 */
  buyerUid?: string;
}) {
  if (!isFirebaseConfigured()) {
    throw new Error("Firebase가 설정되지 않았습니다.");
  }
  try {
    const fn = httpsCallable<
      { productId: string; dealMethod: DealMethod; buyerUid?: string },
      { orderId: string; orderNumber: string }
    >(getClientFunctions(), "createDirectTrade");
    const result = await fn({
      productId: input.productId,
      dealMethod: input.dealMethod,
      ...(input.buyerUid ? { buyerUid: input.buyerUid } : {}),
    });
    if (!result.data?.orderId) throw new Error("거래를 만들지 못했습니다.");
    return result.data;
  } catch (e) {
    throw new Error(stripFirebaseMessage(e));
  }
}

export async function cancelDirectTrade(input: {
  orderId: string;
  cancelReason?: string;
}) {
  if (!isFirebaseConfigured()) {
    throw new Error("Firebase가 설정되지 않았습니다.");
  }
  try {
    const fn = httpsCallable<
      { orderId: string; cancelReason?: string },
      { ok: boolean }
    >(getClientFunctions(), "cancelDirectTrade");
    await fn(input);
  } catch (e) {
    throw new Error(stripFirebaseMessage(e));
  }
}

export async function completeDirectTrade(input: {
  orderId: string;
  fulfillmentMethod: FulfillmentMethod;
  shippingInfo?: TradeShippingInfo | null;
}) {
  if (!isFirebaseConfigured()) {
    throw new Error("Firebase가 설정되지 않았습니다.");
  }
  try {
    const fn = httpsCallable<
      {
        orderId: string;
        fulfillmentMethod: FulfillmentMethod;
        shippingInfo?: TradeShippingInfo | null;
      },
      { ok: boolean }
    >(getClientFunctions(), "completeDirectTrade");
    await fn({
      orderId: input.orderId,
      fulfillmentMethod: input.fulfillmentMethod,
      shippingInfo: input.shippingInfo ?? null,
    });
  } catch (e) {
    throw new Error(stripFirebaseMessage(e));
  }
}

export async function setTradeTransferInfo(input: {
  orderId: string;
  bankName: string;
  accountNumber: string;
  accountHolder: string;
}) {
  if (!isFirebaseConfigured()) {
    throw new Error("Firebase가 설정되지 않았습니다.");
  }
  try {
    const fn = httpsCallable<typeof input, { ok: boolean }>(
      getClientFunctions(),
      "setTradeTransferInfo",
    );
    await fn(input);
  } catch (e) {
    throw new Error(stripFirebaseMessage(e));
  }
}

export async function hideDirectTradeFromHistory(orderId: string) {
  if (!isFirebaseConfigured()) {
    throw new Error("Firebase가 설정되지 않았습니다.");
  }
  try {
    const fn = httpsCallable<{ orderId: string }, { ok: boolean }>(
      getClientFunctions(),
      "hideDirectTradeFromHistory",
    );
    await fn({ orderId });
  } catch (e) {
    throw new Error(stripFirebaseMessage(e));
  }
}
